WordPress/src/main/java/org/wordpress/android/ui/qrcodeauth/QRCodeAuthViewModel.kt

```diff
@@ -12,25 +14,106 @@ import org.wordpress.android.ui.posts.BasicDialogViewModel.DialogInteraction.Neg
+import org.wordpress.android.ui.qrcodeauth.QRCodeAuthUiState.Loading
+import org.wordpress.android.util.NetworkUtilsWrapper
 import javax.inject.Inject
 
+@Suppress("TooManyFunctions")
 @HiltViewModel
-class QRCodeAuthViewModel @Inject constructor() : ViewModel() {
+class QRCodeAuthViewModel @Inject constructor(
+    private val uiStateMapper: QRCodeAuthUiStateMapper,
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    private val validator: QRCodeAuthValidator
+) : ViewModel() {
     private val _actionEvents = Channel<QRCodeAuthActionEvent>(Channel.BUFFERED)
     val actionEvents = _actionEvents.receiveAsFlow()
 
+    private val _uiState = MutableStateFlow<QRCodeAuthUiState>(Loading)
+    val uiState: StateFlow<QRCodeAuthUiState> = _uiState
+
+    private var data: String? = null
+    private var token: String? = null
     private var isStarted = false
 
     fun start() {
         if (isStarted) return
         isStarted = true
+
+        updateUiStateAndLaunchScanner()
+    }
+
+    //  https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=asdfadsfa&data=asdfasdf
+    fun onScanSuccess(scannedValue: String?) {
+        handleScan(scannedValue)
+    }
+
+    fun onScanFailure() {
+        // Note: This is a result of the tap on "X" within the scanner view
+        postActionEvent(FinishActivity)
     }
 
     fun onBackPressed() {
         postActionEvent(LaunchDismissDialog(ShowDismissDialog))
     }
 
+    private fun cancelClicked() {
+        postActionEvent(FinishActivity)
+    }
+
+    private fun authenticateClicked() {
+        // todo: implement
+    }
+
+    private fun handleScan(scannedValue: String?) {
+        extractQueryParamsIfValid(scannedValue)
+
+        if (data.isNullOrEmpty() || token.isNullOrEmpty()) {
+            // todo: handle error
+        } else {
+            postUiState(uiStateMapper.mapLoading())
+            validateScan(data = data.toString(), token = token.toString())
+        }
+    }
+
+    private fun validateScan(data: String, token: String) {
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
+            // todo: handle error
+            return
+        }
+
+        // todo: add authStore.validate and remove below
+        postUiState(
+                uiStateMapper.mapValidated(
+                        "location",
+                        "browser",
+                        this::authenticateClicked,
+                        this::cancelClicked
+                )
+        )
+    }
+
+    private fun extractQueryParamsIfValid(scannedValue: String?) {
+        if (!validator.isValidUri(scannedValue)) return
+
+        val queryParams = validator.extractQueryParams(scannedValue)
+        if (queryParams.containsKey(DATA_KEY) && queryParams.containsKey(TOKEN_KEY)) {
+            this.data = queryParams[DATA_KEY].toString()
+            this.token = queryParams[TOKEN_KEY].toString()
```

Comment:
Are we planning to reset them to null anywhere? Otherwise, the next scan will use previous `data` and `token` values even if those keys are not present in the scanned result.